### ISSUE TITLE
Error: error reading configuration: open PROJECT: no such file or directory

### ISSUE THREAD

**Issue by b1zzu** [operator-framework/operator-sdk#3998]
#### What did you do?

I'm trying to follow the [Generating Manifests and Metadata](https://sdk.operatorframework.io/docs/olm-integration/generation/) tutorial, and I've executed the kustomize generator cmd in an **empty** directory:
```
operator-sdk generate kustomize manifests
```

#### What did you expect to see?

I expected the cmd to ask for the missing information and generate the `kustomization.yaml` and the `.clusterserviceversion.yaml` files

#### What did you see instead? Under which circumstances?

The cmd returned this error:
```
Error: error reading configuration: open PROJECT: no such file or directory
```

#### Environment

**Operator type:**

/language helm

**Kubernetes cluster type:**

`$ operator-sdk version`

```
operator-sdk version: "v1.0.1", commit: "4169b318b578156ed56530f373d328276d040a1b", kubernetes version: "v1.18.2", go version: "go1.13.15 linux/amd64", GOOS: "linux", GOARCH: "amd64"
```

`$ kubectl version`

```
Client Version: version.Info{Major:"1", Minor:"10+", GitVersion:"v1.10.0+b3b92b2", GitCommit:"b3b92b2", GitTreeState:"clean", BuildDate:"2020-07-15T09:27:21Z", GoVersion:"go1.14.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by camilamacedo86** [comment]
Hi @b1zzu, 

Shows that you are trying to use SDK CLI version 1.0+ with a project that did not migrate to the new layout.  You need to migrate your projects see: https://sdk.operatorframework.io/docs/building-operators/helm/migration/

**Comment by camilamacedo86** [comment]
Closing it as solved. However, please feel free to ping us to re-open if you need it.

**Comment by joelanford** [comment]
Re-opening for a bit more discussion because I've seen this question come up a few times.

@camilamacedo86 Can we investigate why that command needs a `PROJECT` file? I wonder if there's a workaround we could suggest or a change we could make to relax this requirement.

**Comment by camilamacedo86** [comment]
Hi @joelanford, 

The PROJECT file is a requirement for the CLI. It has the project config that is used for example to run the commands such as the plugin layout and the project name to integrating the project with OLM.

**Comment by camilamacedo86** [comment]
Hi @joelanford, 

The [PROJECT file](https://github.com/operator-framework/operator-sdk/blob/master/testdata/helm/memcached-operator/PROJECT) is a requirement for the CLI and IMO it should NOT change to keep the consistency across the types as well. It has the project config that is used for example to run the commands such as:

- The project layout: Without it, we do not know if is Helm/Ansible/Go and the plugin version
- The project name: required for the features to integrate the project with OLM.  
- The version (project): to allow what plugins are supported or not and how the tool should work with. See that we can have diff behaviours based on the layout of the PROJECT version as well. 

In the f2f, we spoke about re-create the projects based on the PROJECT file which could bring facilities such as provide a tool to make easier the process to migrate the versions. By storing all data in the PROJECT file it could be easily achieved. 

PS.: Change it would probably mean no longer be able to use the CLI implementation from kb/upstream.

**Comment by joelanford** [comment]
Some SDK subcommands don't require the PROJECT file (e.g. `operator-sdk bundle validate` and `operator-sdk run bundle`). I'm curious what specifically the `generate kustomize manifests` subcommand needs it for?

Is there a workaround we can suggest? (e.g. create a PROJECT file and add some minimal set of info that is required).

One of the goals of operator-sdk v1.0 was to make many of the extra subcommands usable outside the standard project structure, so I'm wondering if we think this command is inextricably tied to the project structure or if it could be run standalone. Looking at the flags exposed by the `generate kustomize manifests` subcommand leaves the impression that it is intended for it to be able to be used in other project structures.

**Comment by camilamacedo86** [comment]
Hi @joelanford, 

IMO it is important to highlight that the root cause is NOT migrated the project. The project file is ONLY the first issue faced as a consequence of it as well.  So, by `I wonder if there's a workaround we could suggest or a change we could make to relax this requirement.` we would end up in another scenario and not really solving it.

**Comment by camilamacedo86** [comment]
Hi @b1zzu,

I got the @joelanford point. If you are looking for only run `operator-sdk generate kustomize manifests` to generate the CSV for your project then the workaround so far would create a file in the root of the project: 

```
$ cat PROJECT 
version: 2
```

Then, it will work. 


However, if you would like to use all commands/features described in the docs you still need to migrate your project for the new layout: https://sdk.operatorframework.io/docs/building-operators/helm/migration/. 

Also, after chatting with @joelanford we might want to facilitate it for no longer be required the workaround so let's use this issue to track this need and see how we can approve it.

**Comment by camilamacedo86** [comment]
Hi @estroz, 

Just to clarifies based on the tests made so far:
- If we add `version: 2` the command is getting the projectName by getting the dir directory name. The same is not true when the version is v3-alpha.
- The resources have been not used in helm to gen the CSV. Should it be used? If yes, then shows that we have a bug. 
- Also, if teh version is v3-alpha than the layout is requested.   

Regards the solution, I have a concern and a point to raise. 

See that if the layout is not in the PROJECT file the stamp used will be : `operators.operatorframework.io.metrics.project_layout: go`

In this way, one of the possible solutions that I'd like to suggest is we ask for the user input the type when the PROJECT file is not found such as we do to ask if the users would like to create the controller and the resource when we run `create api`: 

```sh
$ operator-sdk generate kustomize manifests
Informe your project language [helm/ansible/go/other]
```

**Comment by jmrodri** [comment]
@joelanford one thing I noticed is that just about everything looks at the PROJECT file first, like `operator-sdk version`. The `version` command code does not, I think the cli that configures the plugins is causing this and it is done before all the commands (unless I misread it):

https://github.com/operator-framework/operator-sdk/blob/master/internal/cmd/operator-sdk/cli/cli.go#L50
https://github.com/operator-framework/operator-sdk/blob/master/internal/cmd/operator-sdk/cli/cli.go#L57-L87

I created a `PROJECT` directory in my homedir. Then ran the following:
```
$ operator-sdk version
FATA[0000] failed to read config: read PROJECT: is a directory 
```
version doesn't use anything from PROJECT.

**Comment by joelanford** [comment]
@jmrodri That error is slightly different than the one in this issue.

If you don't have a PROJECT file _at all_, the SDK commands that don't actually require anything from it should work. That's what this issue is about.

If you have a `PROJECT` file, it looks like SDK (via Kubebuilder) expects it to be valid. I don't know if this behavior was intentional or not. If not, we could probably look into ignoring the `PROJECT` file altogether for `ExtraCommands`.

**Comment by jmrodri** [comment]
> @jmrodri That error is slightly different than the one in this issue.
> 
> If you don't have a PROJECT file _at all_, the SDK commands that don't actually require anything from it should work. That's what this issue is about.
> 
> If you have a `PROJECT` file, it looks like SDK (via Kubebuilder) expects it to be valid. I don't know if this behavior was intentional or not. If not, we could probably look into ignoring the `PROJECT` file altogether for `ExtraCommands`.

So running `operator-sdk version` always tries to read the `PROJECT` file if there is one. And if it happens to be a directory which is not a valid PROJECT file, then you get the error I described. And it happens fairly early in the process.

When we call `operator-sdk version` I see the following happening:

operator-sdk.cli.Run() is called  and calls `GetPluginsCLIAndRoot()` https://github.com/operator-framework/operator-sdk/blob/master/internal/cmd/operator-sdk/cli/cli.go#L49-L72
which calls `cli.New`: https://github.com/kubernetes-sigs/kubebuilder/blob/master/pkg/cli/cli.go#L86
which in turn calls `cli.initialize()`: https://github.com/kubernetes-sigs/kubebuilder/blob/master/pkg/cli/cli.go#L99
`cli.initialize()` fails with the following: https://github.com/kubernetes-sigs/kubebuilder/blob/master/pkg/cli/cli.go#L201

That all happens *BEFORE* we call `cli.Run()`: https://github.com/operator-framework/operator-sdk/blob/master/internal/cmd/operator-sdk/cli/cli.go#L51

So I think it's not just "extracommands" that are affected. EVERY command looks at the `PROJECT` file before it's even run.

**Comment by jmrodri** [comment]
The TL;DR is that I believe my comment https://github.com/operator-framework/operator-sdk/issues/3998#issuecomment-705933816 is completely related because it's a command that does NOT need to look at the PROJECT file. Or at least it shouldn't FAIL if it's invalid.

**Comment by joelanford** [comment]
The `kubebuilder` CLI library needs to read the project file to load the plugin which defines the command. So it makes sense for that file to be read for `init` and `create api`, for example.

I agree that we should look to see if we can skip that part for ExtraCommands.

**Comment by estroz** [comment]
@jmrodri can you file a kubebuilder issue for skipping `PROJECT` reads for extra commands, and link it here?

**Comment by jmrodri** [comment]
kubebuilder issue to skip `PROJECT` read for extra commands: https://github.com/kubernetes-sigs/kubebuilder/issues/1755

**Comment by camilamacedo86** [comment]
@jmrodri, 

I was checking this one further and shows that Kubebuilder does not require the PROJECT file for extra commands at all. See that it will not raise an issue when do not exist: https://github.com/kubernetes-sigs/kubebuilder/blob/76e8d21fd0efb89eb7de53bd7b48a1c70c3771d7/pkg/cli/cli.go#L187-L191


The error `Error: error reading configuration: open PROJECT: no such file or directory` faced when we run   `operator-sdk generate kustomize manifests` came from https://github.com/operator-framework/operator-sdk/blob/bbc2f2543770109c438f3953aee95688bdb34bf4/internal/cmd/operator-sdk/generate/kustomize/manifests.go#L97-L100. 

PS.: I am unable to reproduce 
> $ operator-sdk version
> FATA[0000] failed to read config: read PROJECT: is a directory

**Comment by estroz** [comment]
Closed by #4282

**Comment by kyguy** [comment]
Running into the same issue in operator-sdk version: v1.4.0:

When running
```
operator-sdk generate kustomize manifests
```
it returns
```
Error: error reading configuration: open PROJECT: no such file or directory
```
Am I doing something incorrectly? Or was it decided that we should create a PROJECT file in the root dir to get around the error? (e.g 
```echo "version: 2" > PROJECT```). 

**operator-sdk version**
```
operator-sdk version: "v1.4.0", commit: "67f9c8b888887d18cd38bb6fd85cf3cf5b94fd99", kubernetes version: "1.19.4", go version: "go1.15.5", GOOS: "linux", GOARCH: "amd64"
```

**kubectl version**
```
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.0", GitCommit:"af46c47ce925f4c4ad5cc8d1fca46c7b77d13b38", GitTreeState:"clean", BuildDate:"2020-12-08T17:59:43Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by estroz** [comment]
This was supposed to have been fixed by #4074 but on further inspection was not. I'll patch this.

/reopen
/assign

**Comment by estroz** [comment]
On second thought:
- #4074 did not actually intend to fix this problem for `generate kustomize manifests`, just for `generate bundle` and `generate packagemanifests`
- `generate kustomize manifests` probably shouldn't work outside of a project directory for a few reasons:
	1. It was originally intended to work within a project since it reads a bunch of metadata from the PROJECT config
	2. It requires API types for certain project types (Go in particular) be present
	3. The other generate commands now build upon an empty CSV manifest, so you can call them outside of a project if you'd like

I'm going to re-close this since `generate kustomize manifests` is now working as intended. @kyguy feel free to file a feature request if you'd like to see changes to, or have concerns about, what I said above, so discussion can continue.
